pkg/component/pushed_component.go

```diff
@@ -0,0 +1,276 @@
+	return s.dc.Spec.Template.Spec.Containers[0].Env
+}
+
+func (s s2iComponent) GetLabels() map[string]string {
+	return s.dc.Labels
+}
+
+func (s s2iComponent) GetAnnotations() map[string]string {
+	return s.dc.Annotations
+}
+
+func (s s2iComponent) GetName() string {
+	return s.dc.Labels[componentlabels.ComponentLabel]
+}
+
+func (s s2iComponent) GetType() (string, error) {
+	return getType(s)
+}
+
+func (s s2iComponent) GetSource() (string, string, error) {
+	return getSource(s)
+}
+
+type devfileComponent struct {
+	d v1.Deployment
+}
+
+func (d devfileComponent) GetLinkedSecretNames() (secretNames []string) {
+	for _, container := range d.d.Spec.Template.Spec.Containers {
+		for _, env := range container.EnvFrom {
+			if env.SecretRef != nil {
+				secretNames = append(secretNames, env.SecretRef.Name)
+			}
+		}
+	}
+	return secretNames
+}
+
+func (d devfileComponent) GetEnvVars() []v12.EnvVar {
+	var envs []v12.EnvVar
+	for _, container := range d.d.Spec.Template.Spec.Containers {
+		envs = append(envs, container.Env...)
+	}
+	return envs
+}
+
+func (d devfileComponent) GetLabels() map[string]string {
+	return d.d.Labels
+}
+
+func (d devfileComponent) GetAnnotations() map[string]string {
+	return d.d.Annotations
+}
+
+func (d devfileComponent) GetName() string {
+	return d.d.Name
+}
+
+func (d devfileComponent) GetType() (string, error) {
+	return getType(d)
+}
+
+func (d devfileComponent) GetSource() (string, string, error) {
+	return getSource(d)
+}
+
+type noSourceError struct {
+	msg string
+}
+
+func (n noSourceError) Error() string {
+	return n.msg
+}
+
+func getSource(component provider) (string, string, error) {
+	annotations := component.GetAnnotations()
+	if sourceType, ok := annotations[ComponentSourceTypeAnnotation]; ok {
+		if !validateSourceType(sourceType) {
+			return "", "", fmt.Errorf("unsupported component source type %s", sourceType)
+		}
+		var sourcePath string
+		if sourceType == string(config.GIT) {
+			sourcePath = annotations[componentSourceURLAnnotation]
+		}
+
+		klog.V(4).Infof("Source for component %s is %s (%s)", component.GetName(), sourcePath, sourceType)
+		return sourceType, sourcePath, nil
+	}
+	return "", "", noSourceError{msg: fmt.Sprintf("%s component doesn't provide a source type annotation", component.GetName())}
+}
+
+func getType(component provider) (string, error) {
+	if componentType, ok := component.GetLabels()[componentlabels.ComponentTypeLabel]; ok {
+		return componentType, nil
+	}
+	return "", fmt.Errorf("%s component doesn't provide a type label", component.GetName())
+}
+
+func GetPushedComponents(c *occlient.Client, applicationName string) (map[string]PushedComponent, error) {
+	applicationSelector := fmt.Sprintf("%s=%s", applabels.ApplicationLabel, applicationName)
+	dcList, err := c.GetDeploymentConfigsFromSelector(applicationSelector)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			dList, err := c.GetKubeClient().ListDeployments(applicationSelector)
```

Comment:
can we run these two network operations in parallel?

Reply:
We could but I'd rather not complicate things at this point.

Reply:
But for s2i, it will always be one extra unnecessary call before the correct one.